nixos/modules/services/misc/conduit.nix

```diff
@@ -0,0 +1,118 @@
+        description = ''
+          Package of the conduit matrix server to use.
+        '';
+      };
+
+      settings = mkOption {
+        type = format.type;
+        default = {
+        };
+        example = literalExample ''
+          {
+          server_name = "example.com";
+          port = 6167;
+          max_request_size = "20000000";
+          allow_registration = false;
+          allow_encryption = false;
+          allow_federation = false;
+          allow_jaeger = false;
+          trusted_servers = [ "matrix.org" ];
+          address = "127.0.0.1";
+          cache_capacity = 1073741824
+          max_concurrent_requests = 100
+          log = "info,state_res=warn,rocket=off,_=off,sled=off"
+          workers = 4
+          };
+        '';
+        description = ''
+          generates the conduit.toml configuration file. Refer to
+          <link xlink:href="https://gitlab.com/famedly/conduit/-/blob/master/conduit-example.toml"/>
+          for details on supported values.
+          Note that database_path should not be set because the services reliance on systemd StateDir.
+        '';
+      };
+    };
+
+
+    config = mkIf cfg.enable {
+      services.conduit.settings = {
+        database_path = "/var/lib/conduit/conduit.db";
+        port = mkDefault 6167;
+        max_request_size = mkDefault 20000000;
+        allow_registration = mkDefault false;
+        allow_encryption = mkDefault false;
+        allow_federation = mkDefault false;
+        allow_jaeger = mkDefault false;
+        trusted_servers = mkDefault [ "matrix.org" ];
+        address = mkDefault "127.0.0.1";
+      };
+      assertions = [
+        {
+          assertion = cfg.settings.server_name != null;
+          message = "services.conduit.settings.server_name must be set.";
+        }
+        {
+          assertion = (hasAttr "allow_federation" cfg.settings) && cfg.settings.allow_federation -> cfg.iHaveReadTheFederationWarning;
+          message = ''
+            Upstream documentation contains the following warning:
+            You should not join Matrix rooms without asking the admins first.
+            We do not know whether Conduit is safe for general use yet, so you
+            should assume there is some chance that it breaks rooms permanently
+            for all participating users. We are not aware of such a bug today,
+            but we would like to do more testing.
+            You need to acknowledge this warning by setting
+            services.conduit.iHaveReadTheFederationWarning = true
+          '';
+        }
+      ];
+
+      systemd.services.conduit = {
+        description = "Conduit Matrix Server";
+        documentation = [ "https://gitlab.com/famedly/conduit/" ];
+        wantedBy = [ "multi-user.target" ];
+        serviceConfig = {
+          DynamicUser = true;
+          PrivateTmp = true;
+          StateDirectory = "conduit";
+          Environment="CONDUIT_CONFIG=${configFile}";
```

Comment:
Use `systemd.services.conduit.environ` attrset instead. It's cleaner than using strings here.